ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/ParseReportingUrl.java

```diff
@@ -0,0 +1,248 @@
+      ValueProvider<String> countryIpList, ValueProvider<String> osUserAgentList) {
+    return new ParseReportingUrl(urlAllowList, countryIpList, osUserAgentList);
+  }
+
+  private ParseReportingUrl(ValueProvider<String> urlAllowList, ValueProvider<String> countryIpList,
+      ValueProvider<String> osUserAgentList) {
+    this.urlAllowList = urlAllowList;
+    this.countryIpList = countryIpList;
+    this.osUserAgentList = osUserAgentList;
+  }
+
+  private static class InvalidUrlException extends RuntimeException {
+
+    InvalidUrlException(String message) {
+      super(message);
+    }
+
+    public InvalidUrlException(String message, Throwable cause) {
+      super(message, cause);
+    }
+  }
+
+  @Override
+  public Result<PCollection<PubsubMessage>, PubsubMessage> expand(
+      PCollection<PubsubMessage> messages) {
+    return messages.apply(
+        MapElements.into(TypeDescriptor.of(PubsubMessage.class)).via((PubsubMessage message) -> {
+          message = PubsubConstraints.ensureNonNull(message);
+
+          ObjectNode json;
+          try {
+            json = Json.readObjectNode(message.getPayload());
+          } catch (IOException e) {
+            throw new UncheckedIOException(e);
+          }
+
+          try {
+            if (singletonAllowedUrls == null) {
+              loadAllowedUrls();
+            }
+
+            if (singletonCountryToIpMapping == null) {
+              loadCountryToIpMapping();
+            }
+
+            if (singletonOsToUserAgentMapping == null) {
+              loadOsUserAgentMapping();
+            }
+          } catch (IOException e) {
+            throw new UncheckedIOException(e);
+          }
+
+          String reportingUrl = json.path("reporting_url").asText();
+
+          URL urlObj;
+          try {
+            urlObj = new URL(reportingUrl);
+          } catch (MalformedURLException e) {
+            throw new InvalidUrlException("Could not parse reporting URL: " + reportingUrl, e);
+          }
+
+          // TODO: this is placeholder logic
+          // if (urlObj.getHost() == "" || !singletonAllowedUrls.contains(urlObj.getHost())) {
+          // throw new InvalidUrlException("Reporting URL host not found in allow list: "
+          // + reportingUrl);
+          // }
+
+          Map<String, String> attributes = new HashMap<>(message.getAttributeMap());
+
+          // Get query params as map
+          Map<String, String> queryParams = Arrays.stream(urlObj.getQuery().split("&"))
+              .map(param -> param.split("="))
+              .collect(Collectors.toMap(item -> item[0], item -> item[1]));
+
+          String ipParam = singletonCountryToIpMapping
+              .get(attributes.get(Attribute.NORMALIZED_COUNTRY_CODE));
+          if (ipParam == null) {
+            throw new IllegalArgumentException("Could not get ip value: Unrecognized country "
+                + Attribute.NORMALIZED_COUNTRY_CODE);
+          }
```

Comment:
Have we specified that we want to reject messages that don't have known IP or OS values? Will the adM API tolerate having these missing? We may want to check with Nan.